### DISCUSSION TITLE
Help with ics calendar bug?

### DISCUSSION THREAD

**Discussion by ClaireCJS** [pyinstaller/pyinstaller#7674] [Help] [answered]
Okay, this is my 2nd attempt at using PyInstaller, and I'm hoping for it to be my first successful one. 

It sure seems like - this is giving different output when run as a .py than when run as a compiled .exe

I have no idea why.  ChatGPT & I have been going back & forth on this for awhile, but I think it's time to get some actual humans involved.

```
from ics import Calendar as icsCalendar
calendar = icsCalendar(updated_ics_content)                               
print(f"calendar={calendar}")
```

How do I find out why this giving different output after being turned into an EXE?  

For what it's worth - the difference is the same as my difference from GMT.  So it's definitely something to do with timezones being handled differently, somehow, somewhoere.

**Comment by rokm** [comment]
Can you provide a sample for `updated_ics_content`, along with output from .py and from .exe?

Otherwise, you debug this the same way as you would regular python program; check the code of the library you are using, possibly instrumenting it with print statements to see what is going on (both in terms of program flow and the data) and where the difference comes from.

**Reply by rokm** [reply, marked as answer]
If this is about https://github.com/ClioCJS/ingest_ics (protip: if your code is publicly available, do provide the link...) then the problem is likely missing timezone data, indicated by the following message printed by the exe to stderr:
``` 
dateutil\zoneinfo\__init__.py:26: UserWarning: I/O error(2): No such file or directory
``` 

Adding `--collect-data dateutil` to your PyInstaller command (or writing a custom hook for `dateutil` that collects the data using [`PyInstaller.utils.hooks.collect_data_files`](https://pyinstaller.org/en/stable/hooks.html#PyInstaller.utils.hooks.collect_data_files)) should do the trick.

**Reply by ClaireCJS** [reply]
Thank you!  I actually thought I'd included the link, but maybe I got nervous and had erased it.  But yes, it was in that project!

I think I fixed it by adding:

```
    datas=[('C:/ProgramData/anaconda3/Lib/site-packages/dateutil/zoneinfo','dateutil/zoneinfo')],
```

to my spec file.

This is all very new to me and a lot to deal with at once.  I got my program working in an hour, then spent like 6 on the timezone business haha.    I really really want an EXE version. I think now it's working. Thank you!!

**Reply by rokm** [reply]
For the record, the preferred way of dealing with data is to use the afore-mentioned hook utility function, which you can also use directly in the .spec file. At the top of your .spec, add 
```python
from PyInstaller.utils.hooks import collect_data_files
datas = collect_data_files("dateutil")

[...]

a = Analysis(
    ...
    datas=datas,
    ...
)
```

(which is actually what the `--collect-data dateutil` would do for a newly-generated spec).

In this particular case, the hook utility function will collect only `dateutil\zoneinfo\dateutil-zoneinfo.tar.gz` (and will do so without you having to hard-code the path to package, making the spec portable!), whereas adding the directory to `datas` directly will collect everything inside it - including .py files (which you don't need in this case, as byte-compiled modules are already collected into embedded PYZ archive) and `__pycache__` directory.

**Reply by ClaireCJS** [reply]
Oh gosh! That is EXCELLENT advice, and quite perfectly timed! Seriously. Thank you!!

**Reply by ClaireCJS** [reply]
Oh hey, one more question -- if i run the EXE and it's missing the python310.dll, is there a similarly portable solution other than copying my dll in the folder?

**Reply by rokm** [reply]
> Oh hey, one more question -- if i run the EXE and it's missing the python310.dll, is there a similarly portable solution other than copying my dll in the folder?

If the exe is missing the python DLL, then most likely you are trying to run the exe in the `build` directory (which contains intermediate build results) instead of the one in the `dist` directory (which contains actual final frozen application bundle).

**Reply by ClaireCJS** [reply]
> > Oh hey, one more question -- if i run the EXE and it's missing the python310.dll, is there a similarly portable solution other than copying my dll in the folder?
> 
> If the exe is missing the python DLL, then most likely you are trying to run the exe in the `build` directory (which contains intermediate build results) instead of the one in the `dist` directory (which contains actual final frozen application bundle).

OMG. I knew it was dist and not build, yet I completely forgot. OMG. You're really good at this.  Thank you again!

**Comment by ClaireCJS** [comment]
So, I tried your solution in the spec file and ended up with the same timezone offset error. 

But when I did it my way with the unfortunately-hardcoded Anaconda site-lib folder location, I did not end up with the same timezone error.

I wonder what went wrong.

**Reply by rokm** [reply]
What is the output of the following snippet if you run it under your python environment?
```
from PyInstaller.utils.hooks import collect_data_files
datas = collect_data_files("dateutil")
print(datas)
```

**Reply by ClaireCJS** [reply]
Hi, thanks for the tip!   I'm taking notes....

This was my output :)

So just the two entries, but isn't that 2nd one the one in question? hrm....

[code]
Python 3.10.9 | packaged by Anaconda, Inc. | (main, Mar  1 2023, 18:18:15) [MSC v.1916 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
from PyInstaller.utils.hooks import collect_data_files
datas = collect_data_files("dateutil")
print(datas)
[('C:\\ProgramData\\anaconda3\\lib\\site-packages\\dateutil\\zoneinfo\\dateutil-zoneinfo.tar.gz', 'dateutil\\zoneinfo')]
[/code]

**Reply by rokm** [reply]
It's a single entry in form of `(src, dest)` tuple. So the timezone data file should be collected.

If you check your `dist/ingest_ics` directory, does it contain `dateutil\zoneinfo\dateutil-zoneinfo.tar.gz` or not? If it does not (if you have problems with timezones, it likely does not), try rebuilding with added `--clean` argument (`pyinstaller --clean --noconfirm ingest_ics.spec`; and make sure to build with .spec, not the original .py, which would overwrite the spec). 

If that doesn't help, paste the spec file here so we can see if there's something wrong there.

**Reply by ClaireCJS** [reply]
Right now, dateutil-zoneinfo.tar.gz is in the dist/ingest_ics folder, yes.

But I'm kind of scared to change it back to try this out haha.

```
# -*- mode: python ; coding: utf-8 -*-

#still failed: from PyInstaller.utils.hooks import collect_data_files
#still failed: datas = collect_data_files("dateutil")


block_cipher = None


a = Analysis(
    ['ingest_ics.py'],
    pathex=[],
    binaries=[],
    #datas=[],
    datas=[('C:/ProgramData/anaconda3/Lib/site-packages/dateutil/zoneinfo','dateutil/zoneinfo')],
    hiddenimports=[],
    hookspath=[],
    hooksconfig={},
    runtime_hooks=[],
    excludes=[],
    win_no_prefer_redirects=False,
    win_private_assemblies=False,
    cipher=block_cipher,
    noarchive=False,
)
pyz = PYZ(a.pure, a.zipped_data, cipher=block_cipher)

exe = EXE(
    pyz,
    a.scripts,
    [],
    exclude_binaries=True,
    name='ingest_ics',
    debug=False,
    bootloader_ignore_signals=False,
    strip=False,
    upx=True,
    console=True,
    disable_windowed_traceback=False,
    argv_emulation=False,
    target_arch=None,
    codesign_identity=None,
    entitlements_file=None,
)
coll = COLLECT(
    exe,
    a.binaries,
    a.zipfiles,
    a.datas,
    strip=False,
    upx=True,
    upx_exclude=[],
    name='ingest_ics',
)
```

**Reply by rokm** [reply]
> ```
>    #datas=[],
>    datas=[('C:/ProgramData/anaconda3/Lib/site-packages/dateutil/zoneinfo','dateutil/zoneinfo')],
> ```

You need to pass the `datas` list you created at the top of the spec via the keyword `datas` argument: `datas=datas`. See my original example.

**Reply by ClaireCJS** [reply]
> > ```
> >    #datas=[],
> >    datas=[('C:/ProgramData/anaconda3/Lib/site-packages/dateutil/zoneinfo','dateutil/zoneinfo')],
> > ```
> 
> You need to pass the `datas` list you created at the top of the spec via the keyword `datas` argument: `datas=datas`. See my original example.

I'm definitely a bit confused.

If you look, i have 2 lines commented out at the top of my spec file. They are the lines you provided me, and they brought the bug back. When I commented them out and reverted back, the bug disappeared and things were working right.

am I misunderstanding that there was a 3rd option?

**Reply by rokm** [reply]
The correct spec would be:

```
# -*- mode: python ; coding: utf-8 -*-

from PyInstaller.utils.hooks import collect_data_files
datas = collect_data_files("dateutil")


block_cipher = None


a = Analysis(
    ['ingest_ics.py'],
    pathex=[],
    binaries=[],
    datas=datas,  # <---- this is not [] anymore
    hiddenimports=[],
    hookspath=[],
    hooksconfig={},
    runtime_hooks=[],
    excludes=[],
    win_no_prefer_redirects=False,
    win_private_assemblies=False,
    cipher=block_cipher,
    noarchive=False,
)
pyz = PYZ(a.pure, a.zipped_data, cipher=block_cipher)

exe = EXE(
    pyz,
    a.scripts,
    [],
    exclude_binaries=True,
    name='ingest_ics',
    debug=False,
    bootloader_ignore_signals=False,
    strip=False,
    upx=True,
    console=True,
    disable_windowed_traceback=False,
    argv_emulation=False,
    target_arch=None,
    codesign_identity=None,
    entitlements_file=None,
)
coll = COLLECT(
    exe,
    a.binaries,
    a.zipfiles,
    a.datas,
    strip=False,
    upx=True,
    upx_exclude=[],
    name='ingest_ics',
)
```
